### ISSUE TITLE
 Jest SyntaxError: Unexpected token 'export' site:stackoverflow.com

### ISSUE THREAD

**Issue by david-wb** [react-dnd/react-dnd#3443]
After upgrading from `15.1.1` to `16.0.1` I'm getting the following error in my react unit tests. Any help would be much appreciated.

<img width="869" alt="Screen Shot 2022-04-21 at 11 38 35 PM" src="https://user-images.githubusercontent.com/1855225/164610298-78ae524b-7591-45b1-9ae2-4863c4aa7fef.png">

**Comment by eqqe** [comment]
I have react-dnd@16.0.1
and react 18, with CRA, I added the following option to react-scripts (here react-app-rewired)
```react-app-rewired test --transformIgnorePatterns "node_modules/(?!react-dnd)/"```

**Comment by benwiley4000** [comment]
Our team is going to avoid upgrading until this issue is resolved. Seems very unusual to put raw ES exports in the main (non-ES) module.

**Comment by Atlas-1510** [comment]
I am also having this issue:
```
* "@testing-library/jest-dom": "^5.16.4",
* "@testing-library/react": "^13.1.1",
* "react": "^18.0.0",
* "react-dnd": "^16.0.1",
* "react-dnd-html5-backend": "^16.0.1",
```

**Comment by arjandepooter** [comment]
Ran into the same issue, fixed it by making sure `babel-jest` transforms the stuff from `react-dnd`:

`transformIgnorePatterns: ['/node_modules/(?!react-dnd|dnd-core|@react-dnd)']`

As we're using a monorepo with a `babelrc` for each package I added some root config in `babel.config.js` to make sure the node_modules files are transpiled to commonJS modules:

```
  presets: [
    [
      '@babel/preset-env',
      {
        modules: 'commonjs',
        loose: true,
      },
    ],
  ]
```

**Comment by dhirjgupta** [comment]
im using nx monorepo. I tried adding the transformIgnorePatterns in a number of places without luck

**Comment by arjandepooter** [comment]
> im using nx monorepo. I tried adding the transformIgnorePatterns in a number of places without luck

We're also using `nx`. You can add the `transformIgnorePatterns` in your root `jest.preset.js` or your individual jest config files.

**Comment by ozee31** [comment]
Add transformIgnorePatterns not works for me

**Comment by akphi** [comment]
Works for me. Remember that if you already specified another pattern in `transformIgnorePatterns`, there's a risk that it would overlap and cancel the effect of the pattern you added. See https://jestjs.io/docs/configuration#transformignorepatterns-arraystring

For my case

This doesn't work

```js
transformIgnorePatterns: [
    `/node_modules/(?!(somePkg))`,
    `/node_modules/(?!react-dnd|core-dnd|@react-dnd)`,
  ],
```

but this works 

transformIgnorePatterns: [
    `/node_modules/(?!(somePkg)|react-dnd|core-dnd|@react-dnd)`,
  ],
```

**Comment by ozee31** [comment]
does anyone have an example of a complete configuration (jest + babel) that works so I can try to see what's wrong with mine?

**Comment by ozee31** [comment]
Ok it's work, my solution : 

- add `@babel/plugin-transform-modules-commonjs` plugin in `.babelrc`

- in jest.config.js

```js
const config = {
    preset: "ts-jest",
    testEnvironment: "jsdom",
    moduleFileExtensions: ["js", "jsx", "ts", "tsx"],
    testPathIgnorePatterns: ["/node_modules/"],
    testRegex: ".*.(test|spec).(j|t)s[x]?$",
    transform: {
        "node_modules/(react-dnd|dnd-core|@react-dnd)/.+\\.(j|t)sx?$": "ts-jest",
        "^.+\\.js$": "babel-jest",
    },
    transformIgnorePatterns: [`/node_modules/(?!(somePkg)|react-dnd|dnd-core|@react-dnd)`],
};

module.exports = config;
```

**Comment by tyler-hitzeman** [comment]
Anyone else still having issues? Tried what worked for @ozee31 and @arjandepooter, but still getting `Unexpected token "export"` error when importing `useDrag` in jest test.

I have a Lerna monorepo

```javascript
// jest.config.js
...
transformIgnorePatterns: [`/node_modules/(?!(somePkg)|react-dnd|dnd-core|@react-dnd)`],
...
```

```javascript
// babel.config.js <-- just one babel config in monorepo root
...
  presets: [
    [
      '@babel/preset-env',
      {
        modules: 'commonjs',
        loose: true,
      },
    ],
  ]
  
  // also tried replacing the `module` block with the `@babel/plugin-transform-modules-commonjs` plugin
```

**Comment by akphi** [comment]
One way to debug if your `transformIgnorePatterns` is working at all is to go to `node_modules` and put some `console.log` to log out the files being caught by `babel-jest`

**Comment by tyler-hitzeman** [comment]
Thanks @akphi. My issue was that I was adding the `transformIgnorePatterns` to the root of my `jest.config.js`, instead of within the relevant `projects` block. (I split my settings into `web`, `backend` projects)

Here's what worked for me:

```javascript
// jest.config.js

module.exports = {

...

  projects: [
    {
      displayName: "web",
      moduleNameMapper: {
        "^@web/assets/(.*)$": "<rootDir>/packages/web/src/assets/$1",
       ...
      },
      testEnvironment: "jsdom",
      testMatch: ["<rootDir>/packages/web/**/?(*.)+(spec|test).[tj]s?(x)"],
      
     // *** the ignore pattern goes here, within my 'web' project ***
      transformIgnorePatterns: [
        "/node_modules/(?!react-dnd|dnd-core|@react-dnd)",
      ],
    },
    {
      displayName: "backend",
      ...
  }
}
```

I didn't have to make any changes to my babel config, BTW. Here's what that looks like:
```javascript
// babel.config.js
module.exports = {
  presets: [
    "@babel/preset-react",
    ["@babel/preset-env", { targets: { node: "current" } }],
    "@babel/preset-typescript",
  ],
 ...
}

```

**Comment by heniotierra** [comment]
I'm also getting this. Tried the suggested workarounds, but no success for now.

**Comment by 8cells** [comment]
@tyler-hitzeman Thanks, it works

**Comment by EmiM** [comment]
I have the same issue. 
```
jest@26.6.3
react-dnd@16.0.1
```

**Comment by zerico007** [comment]
added this to my package.son and it worked: 

 "jest": {
    "transformIgnorePatterns": [
      "<rootDir>/node_modules/(?!react-dnd|core-dnd|@react-dnd|dnd-core|react-dnd-html5-backend)"
    ]
  }

**Comment by ProdanSergey** [comment]
Nothing here helps to solve it:

jest.config.ts

```
const config: Config.InitialOptions = {
  verbose: true,
  testEnvironment: "jsdom",
  testPathIgnorePatterns: ["/node_modules/"],
  setupFilesAfterEnv: ["<rootDir>/jest-setup.ts"],
  testRegex: ".test.(tsx?)$",
  transform: {
    "^.+\\.tsx?$": "ts-jest",
    "^.+\\.(js|jsx)$": "babel-jest",
  },
  transformIgnorePatterns: ["/node_modules/(?!react-dnd|dnd-core|react-dnd-html5-backend)"],
};
```

tsconfig.json
```
{
  "compilerOptions": {
		"outDir": "./dist",
    "target": "es5",
    "module": "es6",
		"moduleResolution": "Node",
    "strict": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
		"jsx": "react",
		"allowSyntheticDefaultImports": true,
		"types": ["jest"],
		"lib": ["dom", "esnext"],
		"esModuleInterop": true
  },
  "include": ["jest-setup.ts", "./src/**/*"],
  "exclude": ["node_modules"]
}
```
webpack loader

```
{
          test: /\.tsx?$/i,
          use: {
            loader: "ts-loader",
            options: {
              getCustomTransformers: () => ({
                before: [styledComponentsTransformer],
              }),
            },
          },
          exclude: /node_modules/,
        },
```

**Comment by ProdanSergey** [comment]
Here is the approach I've applied to fixed a problem on my side: 

tsconfig.json

```
"compilerOptions": {
    ...
    "allowJs": true
  },
```
jest.config.ts
  
```
  const config: Config.InitialOptions = {
    verbose: true,
    globals: {
      "ts-jest": {
        babelConfig: true,
      },
    },
    preset: "ts-jest",
    testEnvironment: "jsdom",
    testPathIgnorePatterns: ["/node_modules/"],
    setupFilesAfterEnv: ["<rootDir>/jest-setup.ts"],
    testRegex: ".test.(tsx?)$",
    transform: {
      "^.+\\.(j|t)sx?$": "ts-jest",
    },
    transformIgnorePatterns: ["/node_modules/(?!@react-dnd|react-dnd|dnd-core|react-dnd-html5-backend)"],
  };
```

babel.config.js
```
module.exports = {
  presets: ["@babel/preset-env", "@babel/preset-react", "@babel/preset-typescript"],
  plugins: ["@babel/plugin-transform-runtime"],
};
```

**Comment by itsalysialynn** [comment]
> babel.config.js
This was my issue

**Comment by arochadaniel** [comment]
Having this problem as well, and even worse because I'm using SWC + Jest so all the workarounds here don't work. Why including only ESM exports? Why not include CommonJS as well if Jest does not support ESM natively?

Some other packages have this problem with ESM but they include CommonJS exports as well, so I can use a resolver to point to those instead of the ESM ones.

Downgrading to v15 in the meantime

**Comment by sadrataban** [comment]
Adding this to jest.config.js solve it

```
transform: {
        "^.+\\.(j|t)sx?$": "ts-jest",
    },
    transformIgnorePatterns: [
        "/node_modules/(?!@react-dnd|react-dnd|dnd-core|react-dnd-html5-backend)",
    ],
```

**Comment by km-smith** [comment]
I have a TS setup using ts-jest and I attempted @sadrataban setup but it significantly slowed down test runs as I believe its transforming a lot of stuff it doesn't need to be transforming. 

Adding the following to my jest config seemed to do the trick. 

```
transform: {
  "/node_modules/(react-dnd|dnd-core|react-dnd-html5-backend)/dist/(.+).js": "ts-jest",
  "/node_modules/@react-dnd/(invariant|asap|shallowequal)?/dist/(.+).js": "ts-jest",
},
transformIgnorePatterns: [
  "<rootDir>/node_modules/(?!@react-dnd|react-dnd|dnd-core|react-dnd-html5-backend)/",
],
```

The regex can probably be improved but it seemed to speed things up and jest is now happy. Hope it helps someone.

**Comment by lawrence-witt** [comment]
The config I ended up using:

```
const config: Config.InitialOptions = {
  preset: 'ts-jest',
  testEnvironment: 'jsdom',
  transform: {
    'node_modules/(react-dnd|dnd-core|@react-dnd|react-dnd-html5-backend)/.+\\.(j|t)sx?$':
      'ts-jest',
  },
  transformIgnorePatterns: [
    'node_modules/(?!react-dnd|dnd-core|@react-dnd|react-dnd-html5-backend)',
  ],
};
```

It was also necessary to have `"esModuleInterop": true` in tsconfig, otherwise this does nothing.

**Comment by davidslaby** [comment]
@dhirjgupta 
> im using nx monorepo. I tried adding the transformIgnorePatterns in a number of places without luck

Have you resolved the issue in nx mono-repository? :) If you managed, how? Thanks in advance!

**Comment by kanamexx** [comment]
Got settled with followings.
- tsconfig.json
  -  compilerOptions.allowJs: true
  - compilerOptions.esModuleInterop: true
- jest.config.ts
  - transformIgnorePatterns: ["/node_modules/(?!react-dnd|core-dnd|@react-dnd|dnd-core|react-dnd-html5-backend)"]

**Comment by atng** [comment]
@davidslaby @dhirjgupta You can try adding this to your jest.config.js. 

Please note that '/node_modules/' should point to the node_module folder of your monorepo, if react dnd is hoisted. 

```
transformIgnorePatterns: [
    "/node_modules/(?!react-dnd|core-dnd|@react-dnd|dnd-core|react-dnd-html5-backend)",
  ],
```

e.g. if your mono repo is structured as such: -

```
MonoReop
 - node_modules
 - packages
    - react_dnd_app
```
 

then you should add: -

```
transformIgnorePatterns: [
    "../../node_modules/(?!react-dnd|core-dnd|@react-dnd|dnd-core|react-dnd-html5-backend)",
  ],
```

**Comment by ahsanfarooq3423** [comment]
Does someone knows how to use `transformIgnorePatterns` for modules which are not part of node_modules, it is not working for me when I add

```
  testPathIgnorePatterns: ['<rootDir>/src/constants']
```

I want to ignore constants file because it is giving me the following error:

```
    ({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,jest){export * from './toasts';
                                                                                      ^^^^^^

    SyntaxError: Unexpected token 'export'


```

**Comment by john-pimq** [comment]
Set the `transform` and `transformIgnorePatterns` together works for me.

#### jest.config.js
```
transform: {
  'node_modules/(react-dnd|dnd-core|@react-dnd|react-dnd-html5-backend)/.+\\.(j|t)sx?$': "ts-jest"
},
transformIgnorePatterns: [
  'node_modules/(?!(react-dnd|dnd-core|@react-dnd|react-dnd-html5-backend)/)'
]
```

The above code make `ts-jest` to transform `react-dnd` packages to prevent the error of `Unexpected token 'export'`. If someone has more problems with this solution, [v8.0 config with jest](https://github.com/react-dnd/react-dnd/issues/1396) may be useful to reference.

**Comment by surabande** [comment]
React: 18.1.0
react-dnd: 16.0.1

Below changes worked for me.

**Jest.config.ts**

`transformIgnorePatterns: [`/node_modules/(?!${esModules})`]

transform: { 
       [`(${esModules}).+\\.[tj]sx?$`] : ['babel-jest', { presets: ['@nrwl/react/babel'] }]
   }`

**Comment by sascharissling** [comment]
adding this to `package.json` did it for me.
` "jest": {
        "transformIgnorePatterns": [
            "<rootDir>/node_modules/(?!react-dnd|dnd-core|@react-dnd)"
        ]
    }
    `

**Comment by MahyarJ** [comment]
Do we have any plan to fix this specific issue? I keep getting this error even though I did pretty much all workarounds mentioned earlier.

**Comment by binhphan009** [comment]
thanks @akphi 

Your solution saved me.

**Comment by VentuzVictor** [comment]
This [answer](https://github.com/facebook/jest/issues/8006#issuecomment-700755384) fixed the problem in my case, "babel.config.js" was not recognized (it was in the root directory), I have to explicitly say where it is.

By using:

` transform: { '^.+\\.[jt]sx?$': ['babel-jest', { configFile: path.resolve(__dirname, 'babel.config.js') }] },`

**Comment by aninditg98** [comment]
This worked for me

**Comment by cdandrango3** [comment]
type ErrorHandler = (error: mixed, isFatal: boolean) => void;

**Comment by yaKashif** [comment]
Nothing helped. I just stubbed it out for now.

```
jest.mock("react-dnd", () => ({
  useDrag: jest.fn(),
  useDrop: jest.fn(),
  DndProvider: jest.fn(),
}));
jest.mock("react-dnd-html5-backend", () => ({
  HTML5Backend: jest.fn(),
}));
```

**Comment by heyfirst** [comment]
I have this issue and this jest config fix it. I think it's more like jest babel transformation issue? I just don't have time to fix them properly yet.

```
transformIgnorePatterns: ['<rootDir>/node_modules/(?!react-dnd|dnd-core|@react-dnd)'],
```

**Comment by justin-betty** [comment]
Upgraded react from v17 to 18 and jest from 27 to 29 and also running into this issue. Any solutions already?

**Comment by kicks321** [comment]
react-non v16.0.1 is still having this issue:

```ts
SyntaxError: Unexpected token 'export'

  1 | import { useContext, useMemo } from 'react';    
> 2 | import { DndProvider } from 'react-dnd'; 
```

**Comment by kicks321** [comment]
> type ErrorHandler = (error: mixed, isFatal: boolean) => void;

This isn't really a good example of what issues you are to be having, just looks like an implementation of a funcction?

**Comment by bennbollay** [comment]
Attaching a reproduction that doesn't work.  I've tried most every variation above, with the exception of adding babel to the mix, since that seems like it never makes anything better.

Would love for some suggestion as to what to try.  I don't think this is particularly a react-dnd problem, but it is the first one that shows up in Google so, by that standard, it wins.

[example-broken.zip](https://github.com/react-dnd/react-dnd/files/11503624/example-broken.zip)

**Comment by emanuelbodin** [comment]
I got it to work using `"transformIgnorePatterns": ["node_modules/(?!(react-dnd|dnd-core|@react-dnd|react-dnd-html5-backend|react-dnd-touch-backend)/)"]` after renaming my _.babelrc_ to _babel.config.js_. This is important since _.babelrc_ is “File-relative configuration” and _babel.config.*_ is “Project-wide configuration”. Because project-wide config files are separated from the physical location of the config file, they can be ideal for configuration that must apply broadly, even allowing plugins and presets to easily apply to files in `node_modules`.

_babel.config.js_:
`module.exports = {
  presets: ['@babel/preset-env', '@babel/preset-react'],
  plugins: ['@babel/plugin-transform-runtime'],
}
`

_package.json_:
`"jest": {
    "transformIgnorePatterns": [
      "node_modules/(?!(react-dnd|dnd-core|@react-dnd|react-dnd-html5-backend|react-dnd-touch-backend)/)"
    ]
  }`

**Comment by aXtionTim** [comment]
I am using this plugin with NextJS, which automatically disables transpilation of `node_modules` for its Rust compiler. To opt-in to the transpilation, add the packages to the `transpilePackages` property in your `next.config.js`, e.g.:

```js
module.exports = {
    // ... other config options
    transpilePackages: [
      "@react-dnd",
      "dnd-core",
      "react-dnd",
      "react-dnd-html5-backend",
    ],
}
```
Im my case, I had to add more packages because they errored in the jest execution, too. But this is pretty straightforward and clear from the error message.

**Comment by felipfr** [comment]
Thanks @ProdanSergey !!!

**Comment by Haprog** [comment]
Thanks for the tip of using `transformIgnorePatterns`.
I ran into this issue right after starting to use `react-dnd` in a CRA project. I was able to resolve my issue just by adding this into `package.json` and doing no other changes:

```json
  "jest": {
    "transformIgnorePatterns": [
      "[/\\\\]node_modules[/\\\\](?!react-dnd|dnd-core|@react-dnd).+\\.(js|jsx|mjs|cjs|ts|tsx)$",
      "^.+\\.module\\.(css|sass|scss)$"
    ]
  }
```

This is a combination of the [default patterns from CRA](https://github.com/facebook/create-react-app/blob/main/packages/react-scripts/scripts/utils/createJestConfig.js#L51-L54) together with `(?!react-dnd|dnd-core|@react-dnd)` added in to exclude the related packages.

First I tried adding just `"jest": { "transformIgnorePatterns": [ "<rootDir>/node_modules/(?!react-dnd|dnd-core|@react-dnd)" ] }` but then I started getting other errors from elsewhere so I had to combine this with the default ignore pattern from CRA.

**Comment by jdominguezg89** [comment]
> https://github.com/react-dnd/react-dnd/issues/3443#issuecomment-1552301164

This is the only solution that works for me, I tried every single solution posted here, and some others outside this forum. I have a NextJs app and I was importing an external lib btw.

Thanks a lot @bennbollay

**Comment by HristoP96** [comment]
This issue is open for year and a half and its still not resolved....I guess Jest is not a priority for react-dnd team..Pitty

**Comment by g4g1** [comment]
Did you just remove this comment ?

> https://github.com/react-dnd/react-dnd/issues/3443#issuecomment-1764018456

> This issue is open for year and a half and its still not resolved....I guess Jest is not a priority for react-dnd team..Pitty

**Comment by thenano** [comment]
After spending time on this off and on for a couple of years, and never managing to solve with any of the proposed solutions above, we finally had a breakthrough by also finding this issue: https://github.com/facebook/create-react-app/issues/2537

TL;DR: if you have this issue AND you are running a CRA (create-react-app) setup, adding `transformIgnorePatterns` to your package.json (or other jest config files) does not work because CRA hijacks the config. instead, you need to add `--transformIgnorePatterns` to your command line where you run your tests, and all will start working as expected.
Hope this helps other people that are still stuck here!

**Comment by seanconnollydev** [comment]
This is what ended up fixing for me:
```typescript
import type { Config } from "jest";
import nextJest from "next/jest.js";

const createJestConfig = nextJest({
  // Provide the path to your Next.js app to load next.config.js and .env files in your test environment
  dir: "./",
});

// Add any custom config to be passed to Jest
const config: Config = {
  coverageProvider: "v8",
  testEnvironment: "jsdom",
  // Add more setup options before each test is run
  // setupFilesAfterEnv: ['<rootDir>/jest.setup.ts'],
};

// createJestConfig is exported this way to ensure that next/jest can load the Next.js config which is async
export default async () => ({
  ...(await createJestConfig(config)()),
  transformIgnorePatterns: ["node_modules/(?!react-dnd)/"],
});

```

**Comment by davoam** [comment]
In my case changing `transformIgnorePatterns` help  BUT in bunch of answers here `core-dnd` is used while there is `dnd-core` directory (at least in `react-dnd` version `16.0.1`); 

Here is working `transformIgnorePatterns`:
`['/node_modules/(?!react-dnd|dnd-core|@react-dnd|react-dnd-html5-backend)',]`

**Comment by lucas-ennouchi-db** [comment]
> This is what ended up fixing for me:
> 
> ```ts
> import type { Config } from "jest";
> import nextJest from "next/jest.js";
> 
> const createJestConfig = nextJest({
>   // Provide the path to your Next.js app to load next.config.js and .env files in your test environment
>   dir: "./",
> });
> 
> // Add any custom config to be passed to Jest
> const config: Config = {
>   coverageProvider: "v8",
>   testEnvironment: "jsdom",
>   // Add more setup options before each test is run
>   // setupFilesAfterEnv: ['<rootDir>/jest.setup.ts'],
> };
> 
> // createJestConfig is exported this way to ensure that next/jest can load the Next.js config which is async
> export default async () => ({
>   ...(await createJestConfig(config)()),
>   transformIgnorePatterns: ["node_modules/(?!react-dnd)/"],
> });
> ```

Same here with NextJs! thanks!

**Comment by melroy89** [comment]
I believe this should be an upstream issue at Jest. Since Jest doesn't seem to handle code that contain `export type`. Which is just valid TS code.

And yes Jest does support ESM when using `preset: 'ts-jest/presets/default-esm',`
